src/main/java/com/proofpoint/discovery/IpHostnameAuthManager.java

```diff
@@ -0,0 +1,189 @@
+package com.proofpoint.discovery;
+
+import com.google.common.collect.Sets;
+import com.proofpoint.discovery.client.ServiceDescriptor;
+import com.proofpoint.discovery.client.ServiceSelector;
+import com.proofpoint.discovery.store.StoreConfig;
+import com.proofpoint.log.Logger;
+import com.proofpoint.node.NodeInfo;
+import com.proofpoint.units.Duration;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.ForbiddenException;
+import java.net.InetAddress;
+import java.net.URI;
+import java.net.UnknownHostException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+public class IpHostnameAuthManager implements AuthManager
+{
+    private final DynamicStore dynamicStore;
+    private final ServiceSelector discoverySelector;
+    private final String localNodeId;
+    private final Duration updateInterval;
+    private final Set<InetAddress> discoveryAddrs = Sets.newConcurrentHashSet();
+    private Future<?> future;
+    private final ScheduledExecutorService executor;
+
+    private static final Logger logger = Logger.get(IpHostnameAuthManager.class);
+
+    @Inject
+    public IpHostnameAuthManager(DynamicStore dynamicStore,
+            ServiceSelector discoverySelector,
+            NodeInfo nodeInfo,
+            StoreConfig config,
+            @ForAuthManager ScheduledExecutorService executor)
+    {
+        this.dynamicStore = requireNonNull(dynamicStore, "dynamicStore is null");
+        this.discoverySelector = requireNonNull(discoverySelector, "discoverySelector is null");
+        this.localNodeId = requireNonNull(nodeInfo, "nodeInfo is null").getNodeId();
+        this.updateInterval = requireNonNull(config, "config is null").getRemoteUpdateInterval();
+        this.executor = requireNonNull(executor, "executor is null");
+        startDiscoveryRefresh();
+    }
+
+    @PreDestroy
+    public synchronized void preDestroy()
```

Comment:
What is being synchronized against? `@PreDestroy` is guaranteed not to run concurrently with `@PostConstruct` or the constructor.

```suggestion
    private void preDestroy()
```